machxo2/arch.cc

```diff
@@ -0,0 +1,532 @@
+PipId Arch::getPipByName(IdString name) const
+{
+    PipId ret;
+
+    auto it = pip_by_name.find(name);
+    if (it != pip_by_name.end())
+        return it->second;
+
+    Location loc;
+    std::string basename;
+    std::tie(loc.x, loc.y, basename) = split_identifier_name(name.str(this));
+    ret.location = loc;
+
+    const TileTypePOD *tilei = tileInfo(ret);
+    for (int i = 0; i < tilei->num_pips; i++) {
+        PipId curr;
+        curr.location = loc;
+        curr.index = i;
+        pip_by_name[getPipName(curr)] = curr;
+    }
+    if (pip_by_name.find(name) == pip_by_name.end())
+        NPNR_ASSERT_FALSE_STR("no pip named " + name.str(this));
+    return pip_by_name[name];
+}
+
+IdString Arch::getPipName(PipId pip) const
+{
+    NPNR_ASSERT(pip != PipId());
+
+    int x = pip.location.x;
+    int y = pip.location.y;
+
+    std::string src_name = getWireName(getPipSrcWire(pip)).str(this);
+    std::replace(src_name.begin(), src_name.end(), '/', '.');
+
+    std::string dst_name = getWireName(getPipDstWire(pip)).str(this);
+    std::replace(dst_name.begin(), dst_name.end(), '/', '.');
+
+    return id("X" + std::to_string(x) + "/Y" + std::to_string(y) + "/" + src_name + ".->." + dst_name);
+}
+
+// ---------------------------------------------------------------
+
+GroupId Arch::getGroupByName(IdString name) const { return GroupId(); }
+
+IdString Arch::getGroupName(GroupId group) const { return IdString(); }
+
+std::vector<GroupId> Arch::getGroups() const
+{
+    std::vector<GroupId> ret;
+    return ret;
+}
+
+const std::vector<BelId> &Arch::getGroupBels(GroupId group) const { return bel_id_dummy; }
+
+const std::vector<WireId> &Arch::getGroupWires(GroupId group) const { return wire_id_dummy; }
+
+const std::vector<PipId> &Arch::getGroupPips(GroupId group) const { return pip_id_dummy; }
+
+const std::vector<GroupId> &Arch::getGroupGroups(GroupId group) const { return group_id_dummy; }
+
+// ---------------------------------------------------------------
+
+delay_t Arch::estimateDelay(WireId src, WireId dst) const
+{
+    // Taxicab distance multiplied by pipDelay (0.01) and fake wireDelay (0.01).
+    // TODO: This function will not work well for entrance to global routing,
+    // as the entrances are located physically far from the DCCAs.
+    return (abs(dst.location.x - src.location.x) + abs(dst.location.y - src.location.y)) * (0.01 + 0.01);
+}
+
+delay_t Arch::predictDelay(const NetInfo *net_info, const PortRef &sink) const
+{
+    BelId src = net_info->driver.cell->bel;
+    BelId dst = sink.cell->bel;
+
+    NPNR_ASSERT(src != BelId());
+    NPNR_ASSERT(dst != BelId());
+
+    // TODO: Same deal applies here as with estimateDelay.
+    return (abs(dst.location.x - src.location.x) + abs(dst.location.y - src.location.y)) * (0.01 + 0.01);
+}
+
+bool Arch::getBudgetOverride(const NetInfo *net_info, const PortRef &sink, delay_t &budget) const { return false; }
+
+ArcBounds Arch::getRouteBoundingBox(WireId src, WireId dst) const
```

Comment:
This is going to cause `router2` to fail in some exciting way or other, it'd be better to assertion fail than return empty bounds here